File: desktop/renderer/date-entry.test.js
/*
 * `date-entry.js` held to `date-entry-cases.json`, the table `DateEntryTest.kt` reads too, so
 * typing a date works the same way on the phone and on the laptop (#90).
 */

import test from 'node:test';
import assert from 'node:assert';
import { readFileSync } from 'node:fs';

import {
  DateSlot, DateProblem, encode, decode, enter, backspace, settle, problem, isFinal, isDeathBeforeBirth,
} from './date-entry.js';

const table = JSON.parse(readFileSync(new URL('./date-entry-cases.json', import.meta.url)));

const slot = (name) => DateSlot[name.toUpperCase()];

test('the table is the format this test reads', () => {
  assert.strictEqual(table.format, 1);
});

test('every string the slots can hold survives a round trip', () => {
  for (const text of table.roundTrip) {
    assert.strictEqual(encode(decode(text)), text, text);
  }
});

test('typing into a slot fills and moves on as the table says', () => {
  for (const [before, slotName, text, after, focus] of table.enter) {
    const edit = enter(decode(before), slot(slotName), text);
    const name = `${before} + ${slotName}:'${text}'`;
    assert.strictEqual(encode(edit.parts), after, name);
    assert.strictEqual(edit.focus, slot(focus), name);
  }
});

test('backspace in an empty slot steps back through the date', () => {
  for (const [before, slotName, after, focus] of table.backspace) {
    const edit = backspace(decode(before), slot(slotName));
    const name = `${before} ${slotName}`;
    assert.strictEqual(encode(edit.parts), after, name);
    assert.strictEqual(edit.focus, slot(focus), name);
  }
});

test('a lone month or day digit is padded when the date is kept', () => {
  for (const [text, settled] of table.settle) assert.strictEqual(settle(text), settled, text);
});

test('each problem is named and shown only once it is real', () => {
  for (const row of table.problem) {
    const [text, expected, final] = row;
    const found = problem(text);
    assert.strictEqual(found, expected, text);
    if (found != null) assert.strictEqual(isFinal(found, text), final, text);
  }
});

test('only two calendar dates can be out of order', () => {
  for (const [birth, death, expected] of table.deathBeforeBirth) {
    assert.strictEqual(isDeathBeforeBirth(birth, death), expected, `${birth} / ${death}`);
  }
});

test('DateProblem carries the Kotlin enum names, unchanged', () => {
  for (const name of ['YEAR_INCOMPLETE', 'MONTH_OUT_OF_RANGE', 'DAY_OUT_OF_RANGE', 'DAY_NOT_IN_MONTH',
    'NOT_A_LEAP_YEAR', 'DAY_WITHOUT_MONTH', 'MONTH_ALONE', 'MALFORMED', 'DEATH_BEFORE_BIRTH']) {
    assert.strictEqual(DateProblem[name], name);
  }
});

File: app/src/test/java/com/vibethroughcode/ftree/ui/person/DateEntryTest.kt
package com.vibethroughcode.ftree.ui.person

import java.io.File
import kotlinx.serialization.json.Json
import kotlinx.serialization.json.JsonArray
import kotlinx.serialization.json.JsonNull
import kotlinx.serialization.json.boolean
import kotlinx.serialization.json.int
import kotlinx.serialization.json.jsonArray
import kotlinx.serialization.json.jsonObject
import kotlinx.serialization.json.jsonPrimitive
import org.junit.Assert.assertEquals
import org.junit.Test

/**
 * `DateEntry` held to `desktop/renderer/date-entry-cases.json`, the table `date-entry.test.js` reads
 * too, so typing a date works the same way on the phone and on the laptop (#90).
 * `app/build.gradle.kts` declares the table as an input to this task.
 */
class DateEntryTest {

    private val root: File = generateSequence(File(System.getProperty("user.dir")).absoluteFile) { it.parentFile }
        .first { File(it, "settings.gradle.kts").exists() }

    private val table = Json.parseToJsonElement(File(root, "desktop/renderer/date-entry-cases.json").readText()).jsonObject

    private fun rows(key: String): List<List<String?>> = table.getValue(key).jsonArray.map { row ->
        (row as JsonArray).map { if (it is JsonNull) null else it.jsonPrimitive.content }
    }

    private fun slot(name: String?) = DateSlot.valueOf(name!!.uppercase())

    @Test
    fun `the table is the format this test reads`() {
        assertEquals(1, table.getValue("format").jsonPrimitive.int)
    }

    @Test
    fun `every string the slots can hold survives a round trip`() {
        for (text in table.getValue("roundTrip").jsonArray.map { it.jsonPrimitive.content }) {
            assertEquals(text, DateEntry.decode(text).encode())
        }
    }

    @Test
    fun `typing into a slot fills and moves on as the table says`() {
        for ((before, slot, text, after, focus) in rows("enter")) {
            val edit = DateEntry.enter(DateEntry.decode(before!!), slot(slot), text!!)
            val name = "$before + ${slot}:'$text'"
            assertEquals(name, after, edit.parts.encode())
            assertEquals(name, slot(focus), edit.focus)
        }
    }

    @Test
    fun `backspace in an empty slot steps back through the date`() {
        for ((before, slot, after, focus) in rows("backspace")) {
            val edit = DateEntry.backspace(DateEntry.decode(before!!), slot(slot))
            assertEquals("$before $slot", after, edit.parts.encode())
            assertEquals("$before $slot", slot(focus), edit.focus)
        }
    }

    @Test
    fun `a lone month or day digit is padded when the date is kept`() {
        for ((text, settled) in rows("settle")) assertEquals(text, settled, DateEntry.settle(text!!))
    }

    @Test
    fun `each problem is named and shown only once it is real`() {
        for (row in table.getValue("problem").jsonArray.map { it.jsonArray }) {
            val text = row[0].jsonPrimitive.content
            val expected = row[1].let { if (it is JsonNull) null else DateProblem.valueOf(it.jsonPrimitive.content) }
            val problem = DateEntry.problem(text)
            assertEquals(text, expected, problem)
            if (problem != null) assertEquals(text, row[2].jsonPrimitive.boolean, DateEntry.isFinal(problem, text))
        }
    }

    @Test
    fun `only two calendar dates can be out of order`() {
        for (row in table.getValue("deathBeforeBirth").jsonArray.map { it.jsonArray }) {
            val (birth, death) = row.take(2).map { it.jsonPrimitive.content }
            assertEquals("$birth / $death", row[2].jsonPrimitive.boolean, DateEntry.isDeathBeforeBirth(birth, death))
        }
    }
}
